### PR TITLE
update(Link): Get more information as possible from links

### LINKED ISSUE #978: bug(chats): Links from chat conversations are showing an empty icon/preview image
1. Go to a Chat
2. Type a link starting with https:// or http:// (such as https://www.google.com, http://www.satellite.im)
3. Message is sent as a link but the preview icon is empty or showing as a question mark

image

### PR DESCRIPTION
**Pull request by lgmarchi**
### 1. Get more information as possible from links
- Try other ways to get link icon
- Try different ways to get link description 
- It doesn't show Icon, if icon does not exist 

image
image

### Which issue(s) this PR fixes 🔨

- Resolve #978

### CODE REVIEW

**Comment by phillsatellite** [comment]
Can approve once conflicts are gone 💥  Awesome feature bro @lgmarchi

**Review by phillsatellite — APPROVED** [review]

**Review by stavares843 — APPROVED** [review]

File: kit/src/components/embeds/link_embed/get_link_data.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
+use std::error::Error;
+
+use reqwest::Url;
+use scraper::{Html, Selector};
+
+pub async fn get_image_data(document: Html, meta_selector: Selector) -> Option<String> {
```
Comment (sdwoodbury) [inline kit/src/components/embeds/link_embed/get_link_data.rs]: might want to break out  of the loop when the first image is detected. otherwise the last image will be returned.

File: kit/src/components/embeds/link_embed/get_link_data.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            if prop_attr == "og:image" {
+                image = Some(content_attr.to_string());
+            }
+        }
+    }
+
+    image
+}
+
+pub async fn get_title_data(document: Html, meta_selector: Selector) -> String {
```
Comment (sdwoodbury) [inline kit/src/components/embeds/link_embed/get_link_data.rs]: might want to break out of the loop when the first title is detected

File: kit/src/components/embeds/link_embed/get_link_data.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+pub async fn get_image_data(document: Html, meta_selector: Selector) -> Option<String> {
+    let mut image = None;
+
+    for element in document.select(&meta_selector) {
+        let prop_attr_opt = element.value().attr("property");
+        let content_attr_opt = element.value().attr("content");
+
+        if let (Some(prop_attr), Some(content_attr)) = (prop_attr_opt, content_attr_opt) {
+            if prop_attr == "og:image" {
+                image = Some(content_attr.to_string());
```
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/get_link_data.rs:15]: If there are multiple fields with the attribute, are we looking to set the first attribute? If so, we could just return early here.

File: kit/src/components/embeds/link_embed/get_link_data.rs (hunk 2 above)
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/get_link_data.rs]: No need to be async if nothing is awaiting. 

```suggestion
pub fn get_title_data(document: Html, meta_selector: Selector) -> String {
```

File: kit/src/components/embeds/link_embed/mod.rs [hunk 4]
```diff
@@ -17,56 +19,40 @@ pub struct SiteMeta {
...
+                    .unwrap_or_default()
+            } else {
+                data.unwrap_or_default()
+            }
+        }
+        Err(_) => get_image_data(document.clone(), meta_selector.clone())
+            .await
+            .unwrap_or_default(),
+    };
+    let title = get_title_data(document.clone(), meta_selector.clone()).await;
```
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/mod.rs]:
```suggestion
    let title = get_title_data(document.clone(), meta_selector.clone());
```

File: kit/src/components/embeds/link_embed/get_link_data.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    if title.is_empty() {
+        if let Some(element) = document.select(&title_selector).next() {
+            title = element.inner_html();
+        }
+    }
+
+    title
+}
+
+pub async fn get_description_data(document: Html, meta_selector: Selector) -> String {
```
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/get_link_data.rs]: Ditto for not needing an async signature if nothing in the function is awaiting
```suggestion
pub fn get_description_data(document: Html, meta_selector: Selector) -> String {
```

File: kit/src/components/embeds/link_embed/mod.rs [hunk 6]
```diff
@@ -17,56 +19,40 @@ pub struct SiteMeta {
...
+    let description = get_description_data(document.clone(), meta_selector.clone()).await;
```
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/mod.rs]:
```suggestion
    let description = get_description_data(document.clone(), meta_selector.clone());
```

File: kit/src/components/embeds/link_embed/get_link_data.rs (hunk 5 above)
Comment (sdwoodbury) [inline kit/src/components/embeds/link_embed/get_link_data.rs]: same thing about breaking out of the loop

File: kit/src/components/embeds/link_embed/get_link_data.rs (hunk 1 above)
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/get_link_data.rs]:
```suggestion
pub fn get_image_data(document: Html, meta_selector: Selector) -> Option<String> {
```

File: kit/src/components/embeds/link_embed/mod.rs [hunk 7]
```diff
@@ -17,56 +19,40 @@ pub struct SiteMeta {
...
-            };
-            is_desc
-        })
-        .next()
-        .unwrap_or_default();
+    let icon = match fetch_icon(url, document.clone()).await {
+        Ok(data) => {
+            if data.is_none() {
+                get_image_data(document.clone(), meta_selector.clone())
+                    .await
```
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/mod.rs]:
```suggestion
```

File: kit/src/components/embeds/link_embed/mod.rs [hunk 8]
```diff
@@ -17,56 +19,40 @@ pub struct SiteMeta {
...
+                    .unwrap_or_default()
+            } else {
+                data.unwrap_or_default()
+            }
+        }
+        Err(_) => get_image_data(document.clone(), meta_selector.clone())
+            .await
```
Comment (dariusc93) [inline kit/src/components/embeds/link_embed/mod.rs]:
```suggestion
```

File: kit/src/components/embeds/link_embed/mod.rs [hunk 9]
```diff
@@ -17,56 +19,40 @@ pub struct SiteMeta {
...
-                        None
-                    }
-                }
-                None => None,
-            };
-            is_desc
-        })
-        .next()
-        .unwrap_or_default();
+    let icon = match fetch_icon(url, document.clone()).await {
```
Comment (sdwoodbury) [inline kit/src/components/embeds/link_embed/mod.rs]: this code is hard to follow. proposed change: 
```
let icon = if let Ok(Some(d)) = fetch_icon(url, document.clone()).await {
        d
    } else {
        get_image_data(document.clone(), meta_selector.clone())
            .await
            .unwrap_or_default()
    };
    ```

**Review by sdwoodbury — CHANGES_REQUESTED** [review]

**Review by sdwoodbury — APPROVED** [review]

**Review by dariusc93 — APPROVED** [review]
